lib/BloqadeQMC/src/updates/cluster.jl

```diff
@@ -0,0 +1,333 @@
+# Main parts here: huge link_list_update!() and cluster_update!() functions.
+# They don´t seem to call each other -> understand when each is used & how they interact.
+
+function link_list_update!(::AbstractRNG, qmc_state::BinaryQMCState, H::AbstractIsing, ::Diagnostics) # fairly standard, builds necessary data structure for both line and multibranch 
+    Ns = nspins(H)
+    spin_left = qmc_state.left_config
+
+    # retrieve linked list data structures
+    LinkList = qmc_state.linked_list  # needed for cluster update
+    LegType = qmc_state.leg_types
+
+    # A diagonal bond operator has non trivial associates for cluster building
+    Associates = qmc_state.associates
+
+    op_indices = qmc_state.op_indices
+    leg_sites = qmc_state.leg_sites
+
+    if qmc_state isa BinaryGroundState
+        First = qmc_state.first
+
+        # The first N elements of the linked list are the spins of the LHS basis state
+        @inbounds for i in 1:Ns
+            LegType[i] = spin_left[i]
+            Associates[i] = 0
+            First[i] = i
+            if H isa AbstractLTFIM
+                op_indices[i] = 0
+                leg_sites[i] = 0
+            end
+        end
+        idx = Ns
+    else
+        First = fill!(qmc_state.first, 0)  #initialize the First list
+        Last = fill!(qmc_state.last, 0)   #initialize the Last list
+        idx = 0
+    end
+
+    spin_prop = copyto!(qmc_state.propagated_config, spin_left)  # the propagated spin state
+
+    # Now, add the 2M operators to the linked list. Each has either 2 or 4 legs
+    @inbounds for (n, op) in enumerate(qmc_state.operator_list)
+        if issiteoperator(H, op)
+            site = getsite(H, op)
```

Comment:
It would be nice if we can seperate out the code block for each different operator (site and bond)